jaggr-core/src/main/java/com/ibm/jaggr/core/impl/layer/LayerBuilder.java

```diff
@@ -146,58 +169,58 @@ String build() throws IOException {
+					writer.append(notifyLayerListeners(EventType.BEGIN_MODULE, request, entry.getKey()));
 					ModuleInfo info = new ModuleInfo(entry.getKey().getModuleId(), entry.getValue().isScript());
 					LayerContributionType type = (i++ == 0) ? LayerContributionType.BEFORE_FIRST_LAYER_MODULE : LayerContributionType.BEFORE_SUBSEQUENT_LAYER_MODULE;
-					addTransportContribution(sb, type, info);
-					processReader(entry.getValue(), sb);
-					addTransportContribution(sb, LayerContributionType.AFTER_LAYER_MODULE, info);
+					addTransportContribution(type, info);
+					processReader(entry.getKey(), entry.getValue());
+					addTransportContribution(LayerContributionType.AFTER_LAYER_MODULE, info);
 				}
-				addTransportContribution(sb, LayerContributionType.END_LAYER_MODULES, moduleList.getRequiredModules());
+				addTransportContribution(LayerContributionType.END_LAYER_MODULES, moduleList.getRequiredModules());
 			}
 
 			// Now add the loader requested modules
 			if (sorted.getModules().size() > 0) {
-				addTransportContribution(sb, LayerContributionType.BEGIN_MODULES, null);
+				addTransportContribution(LayerContributionType.BEGIN_MODULES, null);
 				int i = 0;
 				for (Map.Entry<IModule, ModuleBuildReader> entry : sorted.getModules().entrySet()) {
-					sb.append(notifyLayerListeners(EventType.BEGIN_MODULE, request, entry.getKey()));
+					writer.append(notifyLayerListeners(EventType.BEGIN_MODULE, request, entry.getKey()));
 					ModuleInfo info = new ModuleInfo(entry.getKey().getModuleId(), entry.getValue().isScript());
 					LayerContributionType type = (i++ == 0) ? LayerContributionType.BEFORE_FIRST_MODULE : LayerContributionType.BEFORE_SUBSEQUENT_MODULE;
-					addTransportContribution(sb, type, info);
-					processReader(entry.getValue(), sb);
-					addTransportContribution(sb, LayerContributionType.AFTER_MODULE, info);
+					addTransportContribution(type, info);
+					processReader(entry.getKey(), entry.getValue());
+					addTransportContribution(LayerContributionType.AFTER_MODULE, info);
 				}
-				addTransportContribution(sb, LayerContributionType.END_MODULES, null);
+				addTransportContribution(LayerContributionType.END_MODULES, null);
 			}
 		}
- 		sb.append(notifyLayerListeners(EventType.END_LAYER, request, null));
-		addTransportContribution(sb, LayerContributionType.END_RESPONSE, null);
+ 		writer.append(notifyLayerListeners(EventType.END_LAYER, request, null));
+		addTransportContribution(LayerContributionType.END_RESPONSE, null);
 
 		moduleList.getDependentFeatures().addAll(dependentFeatures);
 
 		// Output any messages to the console if debug mode is enabled
 		if (options.isDebugMode() || options.isDevelopmentMode()) {
 			for (String errorMsg : errorMessages) {
-				sb.append("\r\nconsole.error(\"" + errorMsg + "\");"); //$NON-NLS-1$ //$NON-NLS-2$
+				writer.append("\r\nconsole.error(\"" + errorMsg + "\");"); //$NON-NLS-1$ //$NON-NLS-2$

```

Comment:
I noticed you changed the StringBuffer to a StringWriter.   I hadn't realized that StringWriter uses a StringBuffer under the hood.  StringBuffer has synchronization characteristics that StringBuilder doesn't.   Do you think it would improve performance if we used StringBuilder here?   I don't believe we share this instance with other threads.

Also specific to this line (and the one a few below): perhaps we should avoid the string plussing.